### PR TITLE
CSHARP-4656 Simplify A : "$A" to A : 1 only on find

### PR DESCRIPTION
**Pull request by sanych-sun**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/LinqProviderAdapterV3.cs [hunk 1]
```diff
@@ -163,7 +163,7 @@ static Expression RemovePossibleConvertToObject(Expression expression)
             var context = TranslationContext.Create(expression, inputSerializer);
             var translation = ExpressionToAggregationExpressionTranslator.TranslateLambdaBody(context, expression, inputSerializer, asRoot: true);
             var (projectStage, projectionSerializer) = ProjectionHelper.CreateProjectStage(translation);
-            var simplifiedProjectStage = AstSimplifier.Simplify(projectStage);
+            var simplifiedProjectStage = AstFindProjectionSimplifier.SimplifyFindProjection(projectStage);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/LinqProviderAdapterV3.cs:166]: I expected this extra step to be in `TranslateExpressionToFindProjection` only.

Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/LinqProviderAdapterV3.cs]: Right, fixed

File: src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Optimizers/AstFindProjectionSimplifier.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿using MongoDB.Driver.Linq.Linq3Implementation.Ast.Expressions;
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Optimizers/AstFindProjectionSimplifier.cs:1]: Add copyright

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
+﻿using System;
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:1]: Add copyright

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        public void Should_project_to_class()
+            => AssertProjection(
+                (Document x) => new Projection{ Text = x.A, Number = x.B },
+                "{ Text : '$A', Number : '$B', _id : 0 }");
+
+        [Fact]
+        public void Should_project_to_anonymous_type()
+            => AssertProjection(
+                (Document x) => new { x.A },
+                "{ A : 1, _id : 0 }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:21]: I would refactor these two tests to be:

    [Fact]
    public void Projection_to_class_should_work()
        => AssertProjection(
            x => new Projection { A = x.A, X = x.B },
            "{ A : 1, X : '$B', _id : 0 }");

    [Fact]
    public void Projection_to_anonymous_type_should_work()
        => AssertProjection(
            x => new { x.A, X = x.B },
            "{ A : 1, X : '$B', _id : 0 }");

(Note shorter property names)

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+        [Fact]
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:23]: I would remove this test

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        public void Should_project_to_anonymous_type_with_custom_fields()
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:24]: Maybe add a test like this also (probably not in this file?):

    [Fact]
    public void Find_projection_should_simplify_as_expected()
    {
        var collection = GetCollection<Document>();
        Expression<Func<Document, Projection>> expression = x => new Projection { A = x.A, X = x.B };
        var projectionDefinition = new ProjectExpressionProjection<Document, Projection>(expression, new ExpressionTranslationOptions());

        var find = collection.Find("{}").Project(projectionDefinition);

        var translatedProjection = TranslateFindProjection(collection, find);
        translatedProjection.Should().Be("{ A : 1, X : '$B', _id : 0 }");

        var results = find.ToList();
    }

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            => AssertProjection(
+                (Document x) => new { A = x.A, Num = x.B },
+                "{ A : 1, Num : '$B', _id : 0 }");
+
+
+        private void AssertProjection<TSource, TProjection>(
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:30]: TSource can just be Document. It doesn't have to be a type parameter.

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+            public string A { get; set; }
+
+            public int B { get; set; }
+        }
+
+        private class Projection
+        {
+            public string Text { get; set; }
+
+            public int Number { get; set; }
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:54]: Probably just A and X for field names. One is the same as Document and the other is not.

**Review by rstam — CHANGES_REQUESTED** [review]
I like your idea of replacing a regular expression projection definition with a `FindExpressionProjectionDefinition` when we know we're doing a `Find`.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Optimizers/AstFindProjectionSimplifier.cs (hunk 2 above)
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Optimizers/AstFindProjectionSimplifier.cs:1]: Done

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs (hunk 3 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:1]: Done

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs (hunk 4 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:21]: Done

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs (hunk 5 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:23]: Done

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs (hunk 7 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:30]: Done

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs (hunk 8 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:54]: Done

File: tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs (hunk 6 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/FindExpressionProjectionDefinitionTests.cs:24]: Added similar test to FindFluentTests without interacting with the database to follow that test class way.

**Review by rstam — CHANGES_REQUESTED** [review]
Let me know if you want to have a Zoom meeting to discuss requested changes.

File: src/MongoDB.Driver/FindFluent.cs [hunk 9]
```diff
@@ -122,6 +122,11 @@ public override Task<long> CountDocumentsAsync(CancellationToken cancellationTok
 
         public override IFindFluent<TDocument, TNewProjection> Project<TNewProjection>(ProjectionDefinition<TDocument, TNewProjection> projection)
         {
+            if (projection is ProjectExpressionProjection<TDocument, TNewProjection> expressionProjection)
+            {
+                projection = new FindExpressionProjectionDefinition<TDocument, TNewProjection>(expressionProjection.Expression);
+            }
+
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/FindFluent.cs:129]: Done, as it was discussed offline via virtual method.

File: tests/MongoDB.Driver.Tests/FindFluentTests.cs [hunk 10]
```diff
@@ -242,6 +242,19 @@ public class FindFluentTests
             }
         }
 
+        [Fact]
+        public void Find_projection_should_simplify_as_expected()
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/FindFluentTests.cs:246]: Done

File: tests/MongoDB.Driver.Tests/FindFluentTests.cs [hunk 11]
```diff
@@ -396,6 +396,12 @@ public class Person
             public string LastName;
             public int Age;
         }
+
+        public class ProjectedPerson
+        {
+            public string FirstName;
+            public string FamilyName;
+        }
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/FindFluentTests.cs:404]: This class is no longer used.

Did you not find a new (better) place for the test you removed in this commit? (The test that used this class)

File: src/MongoDB.Driver/PipelineStageDefinitionBuilder.cs [hunk 12]
```diff
@@ -1960,6 +1960,12 @@ public override RenderedProjectionDefinition<TOutput> Render(IBsonSerializer<TIn
         {
             return linqProvider.GetAdapter().TranslateExpressionToProjection(_expression, inputSerializer, serializerRegistry, _translationOptions);
         }
+
+        internal override RenderedProjectionDefinition<TOutput> RenderForFind(IBsonSerializer<TInput> sourceSerializer, IBsonSerializerRegistry serializerRegistry,
+            LinqProvider linqProvider)
```
Comment (rstam) [inline src/MongoDB.Driver/PipelineStageDefinitionBuilder.cs:1965]: nit: for better readability I prefer that parameters either all be on one line or each on a separate line, as in:

    internal override RenderedProjectionDefinition<TOutput> RenderForFind(
        IBsonSerializer<TInput> sourceSerializer,
        IBsonSerializerRegistry serializerRegistry,
        LinqProvider linqProvider)

File: src/MongoDB.Driver/ProjectionDefinition.cs [hunk 13]
```diff
@@ -147,6 +147,9 @@ public virtual RenderedProjectionDefinition<TProjection> Render(IBsonSerializer<
         /// <returns>A <see cref="RenderedProjectionDefinition{TProjection}"/>.</returns>
         public abstract RenderedProjectionDefinition<TProjection> Render(IBsonSerializer<TSource> sourceSerializer, IBsonSerializerRegistry serializerRegistry, LinqProvider linqProvider);
 
+        internal virtual RenderedProjectionDefinition<TProjection> RenderForFind(IBsonSerializer<TSource> sourceSerializer, IBsonSerializerRegistry serializerRegistry, LinqProvider linqProvider)
+            => Render(sourceSerializer, serializerRegistry, linqProvider);
+
```
Comment (rstam) [inline src/MongoDB.Driver/ProjectionDefinition.cs:152]: This was a really good suggestion to make this method virtual!

We also discussed overriding this method in subclasses where it doesn't make sense and throw `InvalidOperationException`.

**Review by rstam — CHANGES_REQUESTED** [review]
There is still one call to:

    new FindExpressionProjectionDefinition

in `IFindFluentExtensions.cs`.

I think it can be replaced with a call to:

    new ExpressionProjectionDefinition

That would mean that the `FindExpressionProjectionDefinition` class is no longer used at all (and that's a good thing).

We can't really remove the `FindExpressionProjectionDefinition` class because it is `public` and we don't know if any users is creating it directly.

File: tests/MongoDB.Driver.Tests/FindFluentTests.cs (hunk 11 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/FindFluentTests.cs:404]: Removed the unused class. Will create tests once we decide how to share that useful helper methods.

File: src/MongoDB.Driver/PipelineStageDefinitionBuilder.cs (hunk 12 above)
Comment (sanych-sun) [reply src/MongoDB.Driver/PipelineStageDefinitionBuilder.cs:1965]: Done

File: src/MongoDB.Driver/ProjectionDefinition.cs (hunk 13 above)
Comment (sanych-sun) [reply src/MongoDB.Driver/ProjectionDefinition.cs:152]: Done

**Review by rstam — APPROVED** [review]
LGTM

**Comment by connorstorer-kbxcom** [comment]
> There is still one call to:
> 
> ```
> new FindExpressionProjectionDefinition
> ```
> 
> in `IFindFluentExtensions.cs`.
> 
> I think it can be replaced with a call to:
> 
> ```
> new ExpressionProjectionDefinition
> ```
> 
> That would mean that the `FindExpressionProjectionDefinition` class is no longer used at all (and that's a good thing).
> 
> We can't really remove the `FindExpressionProjectionDefinition` class because it is `public` and we don't know if any users is creating it directly.

@rstam Prior to 2.19.2, (2.19.1) this was valid below, where `Projection = Builders<TDocumentEntity>.Projection.Expression` returned FindExpressionProjectionDefinition (public) but now returns sealed `ExpressionProjectionDefinition`. What's your suggestion on how I can continue to Cast/Select/Project this. This is used within a test suite and I know there are a lot of missing pieces below, but I wanted to get your shot from the hip opinion? Much appreciated, no matter how brutal it is :D!

```csharp
public void ApplyProjections<TDocumentEntity, TSelectEntity>()
{
    Expression<Func<TDocumentEntity, TSelectEntity>> select = null;
    var findOpts = new FindOptions<TDocumentEntity, TSelectEntity>
    {
        Projection = Builders<TDocumentEntity>.Projection.Expression(select)
    };
    var projectedData = GetData<TDocumentEntity, TSelectEntity, TSelectEntity>(findOpts.Projection);
}

public async Task<List<TReturnEntity>> GetData<TDocumentEntity, TSelectEntity, TReturnEntity>(
    ProjectionDefinition<TDocumentEntity, TSelectEntity> select)
{
    var queryable = _instance.CollectionDocuments.AsQueryable();
    var additionalExp = GetSomeExpression();
    queryable = queryable.Where(additionalExp);
    var expSelect = select as FindExpressionProjectionDefinition<TDocumentEntity, TSelectEntity>;
    return queryable.Select(expSelect.Expression).Cast<TReturnEntity>().ToList();
}
```

**Comment by connorstorer-kbxcom** [comment]
Bypassing with some temporary reflection as noted in [CSHARP-4860](https://jira.mongodb.org/browse/CSHARP-4860).

**Pull request merged**